src/hotspot/cpu/riscv/riscv_v.ad

```diff
@@ -2951,7 +3844,429 @@ instruct rearrange(vReg dst, vReg src, vReg shuffle) %{
+  effect(TEMP tmp);
+  format %{ "vmask_firsttrue_masked $dst, $src, $v0\t# KILL $tmp" %}
+  ins_encode %{
+    BasicType bt = Matcher::vector_element_basic_type(this, $src);
+    __ vsetvli_helper(bt, Matcher::vector_length(this, $src));
+    __ vmsbf_m(as_VectorRegister($tmp$$reg), as_VectorRegister($src$$reg), Assembler::v0_t);
+    __ vcpop_m($dst$$Register, as_VectorRegister($tmp$$reg));
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+// ------------------------------ Populate Index to a Vector -------------------
+
+instruct populateindex(vReg dst, iRegI src1, immI src2, vReg tmp1, iRegI tmp2) %{
+  match(Set dst (PopulateIndex src1 src2));
+  effect(TEMP_DEF dst, TEMP tmp1, TEMP tmp2);
+  format %{ "populateindex $dst, $src1, $src2\t# KILL $tmp1, $tmp2" %}
+  ins_encode %{
+    BasicType bt = Matcher::vector_element_basic_type(this);
+    Assembler::SEW sew = Assembler::elemtype_to_sew(bt);
+    __ vsetvli_helper(bt, Matcher::vector_length(this));
+    __ vmv_v_x(as_VectorRegister($dst$$reg), as_Register($src1$$reg));
+    __ vid_v(as_VectorRegister($tmp1$$reg));
+    __ li(as_Register($tmp2$$reg), (int)($src2$$constant));
+    __ vmacc_vx(as_VectorRegister($dst$$reg), as_Register($tmp2$$reg), as_VectorRegister($tmp1$$reg));
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct insertI_index_lt32(vReg dst, vReg src, iRegI val, immI idx, vRegMask_V0 v0) %{
+  predicate(n->in(2)->get_int() < 32 &&
+            (Matcher::vector_element_basic_type(n) == T_BYTE ||
+             Matcher::vector_element_basic_type(n) == T_SHORT ||
+             Matcher::vector_element_basic_type(n) == T_INT));
+  match(Set dst (VectorInsert (Binary src val) idx));
+  effect(TEMP v0);
+  format %{ "insertI_index_lt32 $dst, $src, $val, $idx" %}
+  ins_encode %{
+    BasicType bt = Matcher::vector_element_basic_type(this);
+    __ vsetvli_helper(bt, Matcher::vector_length(this));
+    __ vid_v(as_VectorRegister($v0$$reg));
+    __ vadd_vi(as_VectorRegister($v0$$reg), as_VectorRegister($v0$$reg), -16);
+    __ vmseq_vi(as_VectorRegister($v0$$reg), as_VectorRegister($v0$$reg), (int)($idx$$constant) - 16);
+    __ vmerge_vxm(as_VectorRegister($dst$$reg), as_VectorRegister($src$$reg), $val$$Register);
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct insertI_index(vReg dst, vReg src, iRegI val, immI idx, iRegINoSp tmp1, vReg tmp2, vRegMask_V0 v0) %{
+  predicate(n->in(2)->get_int() >= 32 &&
+            (Matcher::vector_element_basic_type(n) == T_BYTE ||
+             Matcher::vector_element_basic_type(n) == T_SHORT ||
+             Matcher::vector_element_basic_type(n) == T_INT));
+  match(Set dst (VectorInsert (Binary src val) idx));
+  effect(TEMP tmp1, TEMP tmp2, TEMP v0);
+  format %{ "insertI_index $dst, $src, $val, $idx\t# KILL $tmp1, $tmp2" %}
+  ins_encode %{
+    BasicType bt = Matcher::vector_element_basic_type(this);
+    __ vsetvli_helper(bt, Matcher::vector_length(this));
+    __ vid_v(as_VectorRegister($v0$$reg));
+    __ mv($tmp1$$Register, (int)($idx$$constant));
```

Comment:
Suggestion: make `idx` an register input operand and eliminate this `mv` instruction and the `tmp1` reserved.

Reply:
Thanks for the review! Fixed.